File: coding_exercise/count_ones.py


def count_ones(input_num):
    '''
    Counting # of digit one shown up. NOTE: this is a math problem :(
    '''
    if input_num <= 0:
        return 0
    if input_num < 10:
        return 1
    res = 0
    val = input_num
    div = 1
    # Begin to calculate ones digit by digit. Starting from LSB
    while val:
        current_digit = val % 10
        val /= 10
        if current_digit == 0:
            # Special case - ones are with MSB from 1 - (MSB - 1)
            res += (val) * div
        elif current_digit == 1:
            # Special case - ones are with MSB from 0 - (MSB - 1)
            # and modulo of division
            res += (val) * div
            res += (input_num % div) + 1
        else:
            # Normal case - ones are with MSB from 0 - MSB
            res += (val + 1) * div
        div *= 10
    return res


def test(in_val):
    print "Input: %d" % (in_val)
    print "My answer: %d" % (count_ones(in_val))


test(-1)
test(0)
test(19)
test(99)
test(109)
test(119)
test(209)
test(319)
test(999)
test(5746)

File: coding_exercise/count_ones.cpp
#include "basic_includes.hpp"

using namespace std;


class Solution {
    public:
        int countDigitOne(int n)
        {
            if (n <= 0) return 0;
            if (n < 10) return 1;
            int value = n;
            int div = 1;
            int sum = 0;
            while (value)
            {
                int current_digit = value % 10;
                value /= 10;
                if (current_digit == 0)
                {
                    sum += value * div;
                }
                else if (current_digit == 1)
                {
                    sum += value * div;
                    sum += n % div + 1;
                }
                else
                {
                    sum += (value + 1) * div;
                }
                div *= 10;
            }
            return sum;
        }
};

void test(int input)
{
    Solution sol;
    cout << "Input: " << input << " " << sol.countDigitOne(input) << endl;
}

int main()
{
    test(-1);
    test(0);
    test(19);
    test(99);
    test(109);
    test(119);
    test(209);
    test(319);
    test(999);
    test(5746);
    return 0;
}
